[AArch64] Add a check for invalid default features

tommat01 opened the pull request:
This adds a check that all ExtensionWithMArch which are marked as implied features for an architecture are also present in the list of default features. It doesn't make sense to have something mandatory but not on by default.

There were a number of existing cases that violated this rule, and some changes to which features are mandatory (indicated by the Implies field).

This resulted in a bug where if a feature was marked as `Implies` but was not added to `DefaultExt`, then for `-march=base_arch+nofeat` the Driver would consider `feat` to have never been added and therefore would do nothing to disable it (no `-target-feature -feat` would be added, but the backend would enable the feature by default because of `Implies`). See [clang/test/Driver/aarch64-negative-modifiers-for-default-features.c.](https://github.com/llvm/llvm-project/pull/104435/files#diff-e355e3951d191d3a32265d9bdeb101e4f49ddfa6049ef058cf9e1dfdf7c19ef3)

Note that the processor definitions do not respect the architecture DefaultExts. These apply only when specifying `-march=<some architecture version>`. So when a feature is moved from `Implies` to `DefaultExts` on the Architecture definition, the feature needs to be added to all processor definitions (that are based on that architecture) in order to preserve the existing behaviour. I have checked the TRMs for many cases (see specific commit messages) but in other cases I have just kept the current behaviour and not tried to fix it.

> FEAT_SPECRES is mandatory from Armv8.5.

Added to the HasV8_5aOps.DefaultExts.

> FEAT_DIT is mandatory from Armv8.4.

Added to the HasV8_4aOps.DefaultExts.

> FEAT_SSBS is not mandatory for any architecture.

https://reviews.llvm.org/D54629 says it is mandatory for 8.5-a but I can't see that in the Arm ARM.
Removed from Implied and added to HasV8_5aOps.DefaultExts.

Added to Apple A15/A16/A17 and (presumably?) M4 processor features.
https://developer.apple.com/download/apple-silicon-cpu-optimization-guide/

> FEAT_BTI is mandatory from Armv8.5.

FeatureBranchTargetId is added to the DefaultExts

> FEAT_FlagM is mandatory from Armv8.4.

FeatureFlagM is added to the DefaultExts

> In an Armv8.4 implementation, if FEAT_AdvSIMD is implemented, FEAT_DotProd is implemented.

Added to the HasV8_4aOps.DefaultExts.

> FEAT_SB is mandatory from Armv8.5.

Added to HasV8_5aOps.DefaultExts.
Therefore it appears on the `-cc1` command line for `-march=armv8.5a+nosb`.

> FEAT_WFxT is mandatory from Armv8.7.

Added to HasV8_7aOps.DefaultExts and HasV9_2aOps.DefaultExts.

> FEAT_CCIDX is OPTIONAL from Armv8.2.

Removed from Implies and added to HasV8_3aOps.DefaultExts and HasV8_0rOps.DefaultExts.
This caused it to be removed from many cores.
- Added back for the relevant Cortex/Neoverse cores.
- Removed from Apple cores.
- Added to Ampere/Saphira/ThunderX3T110 to preserve behaviour (I don't know whether they support it)

For v8-R, FEAT_CCIDX is not mandatory, removed.
 - Not implemented by cortex-r82, removed: https://developer.arm.com/documentation/102670/0300/?lang=en
 - Ditto cortex-r82ae

tommat01 committed:
[AArch64] Add a check for invalid default features

This adds a check that all ExtensionWithMArch which are marked as
implied features for an architecture are also present in the list of
default features. It doesn't make sense to have something mandatory but
not on by default.

There were a number of existing cases that violated this rule, and some
changes to which features are mandatory (indicated by the Implies
field):

> FEAT_SPECRES is mandatory from Armv8.5.
FeaturePredRes is added to the HasV8_5aOps.DefaultExts.

> FEAT_DIT is mandatory from Armv8.4.
FeatureDIT is added to the HasV8_4aOps.DefaultExts.

FEAT_SSBS is not mandatory for any architecture.
https://reviews.llvm.org/D54629 says it is mandatory for 8.5-a but I can't see that in the Arm ARM.
FeatureSSBS is removed from Implied and added to HasV8_5aOps.DefaultExts.
Cortex-A710 does not appear to have SSBS
https://developer.arm.com/documentation/101800/0201/The-Cortex-A710--core/Cortex-A710--core-features?lang=en
Removed from the Cortex-A710 and Oryon print-supported-extensions tests.
https://developer.apple.com/download/apple-silicon-cpu-optimization-guide/
Added to Apple A15/A16/A17 and (presumably?) M4 processor features.

> FEAT_BTI is mandatory from Armv8.5.
FeatureBranchTargetId is added to the DefaultExts

> FEAT_FlagM is mandatory from Armv8.4.
FeatureFlagM is added to the DefaultExts

> In an Armv8.4 implementation, if FEAT_AdvSIMD is implemented, FEAT_DotProd is implemented.
FeatureDotProd is added to the HasV8_4aOps.DefaultExts.
FIXME what about nofp here?

> FEAT_SB is mandatory from Armv8.5.
FeatureSB is added to HasV8_5aOps.DefaultExts.
Therefore it appears on the `-cc1` command line for `-march=armv8.5a+nosb`.

> FEAT_WFxT is mandatory from Armv8.7.
FeatureWFxT is added to HasV8_7aOps.DefaultExts and HasV9_2aOps.DefaultExts.

> FEAT_CCIDX is OPTIONAL from Armv8.2.
Removed from Implies and added to HasV8_3aOps.DefaultExts and HasV8_0rOps.DefaultExts.

For v8-R, FEAT_CCIDX is not mandatory, removed.
 - Not implemented by cortex-r82, removed:
   https://developer.arm.com/documentation/102670/0300/?lang=en
 - Ditto cortex-r82ae

davemgreen commented:
> Cortex-A710 does not appear to have SSBS

It would be very surprising to drop it from one generation of cpus, only to add it again in the next!

I believe this says it should be present:
https://developer.arm.com/documentation/101800/0201/AArch64-registers/AArch64-Identification-register-summary/ID-PFR2-EL1--AArch32-Processor-Feature-Register-2?lang=en
ID_AA64PFR1_EL1 also has a reference to it.

tmatheson-arm commented:
> > Cortex-A710 does not appear to have SSBS
> I believe this says it should be present: 

You're right, I'll fix that.

I will also actually remove FeatureCCIDX from the 8.3 mandatory features.

jthackray reviewed (commented):
It looks like you've still got the text "FIXME what about nofp here?" in the commit message.

pratlucas commented on llvm/utils/TableGen/ARMTargetDefEmitter.cpp:
```diff
@@ -19,10 +19,38 @@
+  std::set<Record *> SeenFeats;
+  CollectImpliedFeatures(SeenFeats, Root);
+
+  // For processors, check that each of the mandatory (implied) features which
```
Nit: this seems to be covering architecture versions rather than processors.

jthackray reviewed (approved):
C++ code looks good. Does this also fix the "+nossbs" issue we saw earlier this week?

Presumably this should be cherry-picked to the llvm19 branch?

tommat01 committed:
Remove CCIDX from defaults and from Apple cores

tommat01 committed:
Add CCIDX back to Ampere cores.

I don't know if this is correct but it preserves the existing behaviour.

tommat01 committed:
Add CCIDX back to Cortex A510/520/520AE

https://developer.arm.com/documentation/101604/latest/
https://developer.arm.com/documentation/102517/0003/The-Cortex-A520--core/Supported-standards-and-specifications?lang=en
https://developer.arm.com/documentation/107726/0000/The-Cortex-A520AE--core/Supported-standards-and-specifications?lang=en

tommat01 committed:
Cortex A520AE CCIDX

https://developer.arm.com/documentation/107726/0000/The-Cortex-A520AE--core/Supported-standards-and-specifications?lang=en

tommat01 committed:
FeatureCCIDX, FeatureSSBS for Cortex A7*

tommat01 committed:
CCIDX for Cortex X2/3/4/925

https://developer.arm.com/documentation/101803/0201/The-Cortex-X2--core/Supported-standards-and-specifications?lang=en
https://developer.arm.com/documentation/102807/0001/The-Cortex-X925--core/Supported-standards-and-specifications?lang=en

tommat01 committed:
CCIDX for neoverse cores

https://developer.arm.com/documentation/102099/0003/The-Neoverse-N2--core/Supported-standards-and-specifications?lang=en
https://developer.arm.com/documentation/101427/0102/?lang=en

tommat01 committed:
Add CCIDX back to Saphira and ThunderX3T110

I don't know if this is correct but it preserves the existing behaviour.

tommat01 committed:
Add a test for the -march bug disabling default extensions

tommat01 committed:
fix comment

tmatheson-arm commented:
Fixed SSBS and CCIDX.

> Does this also fix the "+nossbs" issue we saw earlier this week?

Yes, added a test https://github.com/llvm/llvm-project/pull/104435/files#diff-e355e3951d191d3a32265d9bdeb101e4f49ddfa6049ef058cf9e1dfdf7c19ef3

tommat01 committed:
Add SSBS back to Oryon

tmatheson-arm commented on llvm/utils/TableGen/ARMTargetDefEmitter.cpp:
```diff
@@ -283,9 +311,7 @@ static void EmitARMTargetDef(RecordKeeper &RK, raw_ostream &OS) {
 
-    // The apple-latest alias is backend only, do not expose it to -mcpu.
-    if (Name == "apple-latest")
-      continue;
```
This check is duplicated at line 260, unrelated to this change.

tommat01 committed:
Merge remote-tracking branch 'upstream/main' into default_features_fix

pratlucas reviewed (approved):
LGTM.

tommat01 committed:
fix TargetParserTests

tommat01 committed:
fix SSBS test

tommat01 committed:
fix CCIDX tests

tommat01 committed:
clang-format

tommat01 committed:
More clang-format

tmatheson-arm commented:
@jroelofs @ahmedbougacha Are you happy with the FEAT_CCIDX changes to Apple processors?

ahmedbougacha commented on llvm/lib/Target/AArch64/AArch64Processors.td:
```diff
@@ -845,7 +857,8 @@ def ProcessorFeatures {
                                      FeatureComplxNum, FeatureCRC, FeatureJS, FeatureLSE,
-                                     FeaturePAuth, FeatureRAS, FeatureRCPC, FeatureRDM];
+                                     FeaturePAuth, FeatureRAS, FeatureRCPC, FeatureRDM,
+                                     FeatureCCIDX];
```
CCIDX should be removed in this one

ahmedbougacha commented on llvm/lib/Target/AArch64/AArch64Processors.td:
```diff
@@ -892,7 +905,7 @@ def ProcessorFeatures {
                                     FeatureLSE, FeaturePAuth, FeatureFPAC,
                                     FeatureRAS, FeatureRCPC, FeatureRDM,
-                                    FeatureDotProd, FeatureMatMulInt8];
+                                    FeatureDotProd, FeatureMatMulInt8, FeatureSSBS];
```
AppleM4 actually does not have SSBS, unlike the previous ones

ahmedbougacha reviewed (commented):
A couple small issues inline, LGTM otherwise, thanks!

>> FEAT_SSBS is not mandatory for any architecture.
>
> https://reviews.llvm.org/D54629 says it is mandatory for 8.5-a but I can't see that in the Arm ARM.

It was indeed originally mandatory (see e.g., DDI 0487E.a) but it isn't anymore (allowing it to be removed from later cores)

tommat01 committed:
Fix apple-a12 and apple-m4

tommat01 committed:
Merge remote-tracking branch 'upstream/main' into default_features_fix

tommat01 committed:
test changes

tmatheson-arm commented:
/cherry-pick 362142c4bb5cc657151f592e507f552d5b9f7dde

lenary commented:
Just closing a loop because I didn't have the info until today: the Oryon info in this patch is correct, the core does have SSBS and CCIDX.
